src/i2c.rs

```diff
@@ -767,6 +771,53 @@ where
         self.i2c.cr1.modify(|_, w| w.pe().set_bit());
     }
 
+    fn check_and_clear_error_flags(&self) -> Result<(), Error> {
+        let sr1 = self.i2c.sr1.read();
+
+        // Clear all pending error flags. We have already read the SR1, so it's safe to clear them
+        // before returning the error code.
+        self.i2c.sr1.modify(|_, w| {
```

Comment:
It appears that modifying this register before checking the bits below causes the below checks to never succeed.